### DISCUSSION TITLE
Remove an inherited field from a model in pydantic v2

### DISCUSSION THREAD

**Discussion by Top-Dog** [pydantic/pydantic#6699] [Pydantic V2]
With pydantic v1 it was possible to exclude named fields in the child model if they were inherited from the parent with: 
```py
class Config:
        fields = {'clinic_id': {'exclude': True}}
```
The fields member variable has been removed in pydantic v2, so how can the same functionality be achieved? As a follow up question, what if the field is not inherited but part of a complex type in a model, for example, how to remove the id_ field from the following model in only one use case (CreatTicketOut):

```py
class TicketCreate(BaseModel):
    clinic_id: int = Field(default=..., ge=0, description='Customer ID', example=24)
    text: str = Field(default=..., max_length=5000, description='Issuer', example='')
    id_: int # Internal ID for Database - I don't want to use Field(exclude=True, title="DB id") here because other models will inherit from TicketCreate and should include id_ and have access to it)

class CreatTicketOut(BaseModel):
    create_ticket: TicketCreate
    # Exclude only id_ in create_ticket?
```

See previous discussion of this topic for pydantic v1: https://github.com/pydantic/pydantic/discussions/2686

**Comment by xtendch** [comment]
I am struggling with the same question. Currently my approach is to annotate attributes in inherited models as ClassVar as they are then automatically excluded. see https://docs.pydantic.dev/latest/usage/models/#automatically-excluded-attributes

But I am not sure what the implications and drawbacks of this approach are. Is it safe? Is it the best way forward?

Your example could then look like:
```py
class TicketCreate(BaseModel):
    clinic_id: int = Field(default=..., ge=0, description='Customer ID', example=24)
    text: str = Field(default=..., max_length=5000, description='Issuer', example='')
    id_: int # Internal ID for Database - I don't want to use Field(exclude=True, title="DB id") here because other models will inherit from TicketCreate and should include id_ and have access to it)

class CreatTicketOut(TicketCreate):
    id_: ClassVar[int] # Exclude only id_ from field list of model instances
```

I love some of Pydantic V2's new features (e.g. `@computed_field`) but I am struggling in finding a convenient way to replace the exclude/include config for inherited models.

**Reply by alexferl** [reply]
This works and it also removes them from the FastAPI schema which is what I needed. Thanks!

**Comment by sydney-runkle** [comment]
Hi, thanks for your question.

See [this documentation](https://docs.pydantic.dev/latest/usage/serialization/#advanced-include-and-exclude) for how to use the `include` and `exclude` arguments of the `model_dump` function to achieve what you're looking for. This documentation not only includes how to exclude fields of nested fields, as well as fields which take the form of lists or tuples.

Here's a basic example:
```py
from pydantic import BaseModel
from typing import List


class Credentials(BaseModel):
    username: str
    password: str


class User(BaseModel):
    name: str
    credentials: Credentials

user = User(name="charles", credentials=Credentials(username="charles123", password="super-secret-pw"))

print(user.model_dump())
# > {'name': 'charles', 'credentials': {'username': 'charles123', 'password': 'super-secret-pw'}}
print(user.model_dump(exclude={"credentials": {"password"}}))
# > {'name': 'charles', 'credentials': {'username': 'charles123'}}
```

Feel free to let me know if you have any follow up questions!

**Reply by xtendch** [reply]
Thanks for your reply. But this is not really an option as it would required to make sure we exclude the credentials wherever we use the User model. For security reasons I would like to re-use a model as it was possible with pydantic V1.
E.g. 
```py
from pydantic import BaseModel
from typing import List


class Credentials(BaseModel):
    username: str
    password: str


class User(BaseModel):
    name: str
    credentials: Credentials

class UserOut(User):
    ...
    class Config:
        fields = {
            "credentials": {"exclude": True},
        }
```
Is there no efficient code-saving way to achieve this anymore?

**Reply by calmacx** [reply]
@xtendch did you get anywhere with this? I'm having the exact same issue

**Reply by xtendch** [reply]
Unfortunately I haven't found a good solution yet. For the meantime we are not switching to pydantic v2 as this point is very crucial and we made extensive use of the exclude.

**Reply by calmacx** [reply]
I found a pretty horrible hacky temporary solution:
```py
class UserOut(BaseModel):
    ....

if type(UserOut.__fields_set__) == set:
      UserOut.__fields_set__.remove('credentials')
del UserOut.model_fields['credentials']
```

Doubt it'd be recommended/safe to do.. though is working for me, for now. I too would much prefer to be able to use "exclude" as in v1

**Reply by xtendch** [reply]
https://github.com/pydantic/pydantic/discussions/2686 seems to be dealing with that question too. (But without an answer that sounds good to me.)

**Comment by xtendch** [comment]
Can anyone suggest a safe and good way to do this?

**Comment by xtendch** [comment]
I still don't have a satisfying answer to this problem, but omitting fields as suggested in https://github.com/pydantic/pydantic/discussions/8516 could solve this issue too.

**Reply by H-G-11** [reply]
I think I found an easier answer below ⬇️

**Comment by H-G-11** [comment]
You can redefine the field in the child model, but with `Field(..., exclude=True)`

Here is the complete code:

```py
from pydantic import Field, BaseModel

class Parent(BaseModel):
    a: int = 1

class Child(Parent):
    a: int = Field(default=1, exclude=True)

print(Parent(a=2).model_dump())
# returns {'a': 2}
print(Child().model_dump())
# returns {}
```

Works with pydantic v2.5.0

**Reply by bayangan1991** [reply]
Works to a point but unfortunately doesn't remove it from the schema or instance

```python
>>> Child.model_json_schema()
{'properties': {'a': {'default': 1, 'title': 'A', 'type': 'integer'}}, 'title': 'Child', 'type': 'object'}
>>> Child()
a=1
```

**Reply by H-G-11** [reply]
I agree, but I don't think it has an impact for most use-cases, as in the end, what we mostly do is putting it in JSON format and uploading it somewhere.

Not sure what FastAPI would return in an endpoint returning `Child()`

**Reply by bayangan1991** [reply]
All my uses cases unfortunately use the schema :(

Can't generate reliable OpenAPI SDKs if the schema doesn't properly represent the object I need.

Back to handcrafting I guess.

**Reply by H-G-11** [reply]
Well, if the schema is the only issue, I guess [this](https://docs.pydantic.dev/latest/api/json_schema/#pydantic.json_schema.SkipJsonSchema) should work


```py
from pydantic import Field, BaseModel
from pydantic.json_schema import SkipJsonSchema

class Parent(BaseModel):
    a: int = 1

class Child(Parent):
    a: SkipJsonSchema[int] = Field(default=1, exclude=True)

print(Parent(a=2).model_dump())
print(Child().model_dump())
print(Child.model_json_schema())
```

Outputs: 
```py
{'a': 2}
{}
{'properties': {}, 'title': 'Child', 'type': 'object'}
```

Having this kind of extra hidden field is not ideal, and am not a big fan of it neither, but that's a workaround

**Reply by seankerr** [reply]
Appreciate your resolution on this @H-G-11. This should be in the official documentation.

**Comment by xtendch** [comment]
ChatGPT brought me on the idea of using the @model_serialize decorator. Is there anything speaking against using this approach? Is using model_copy not ideal?

Example with include (which is nice, because you don't have to remember about excluding potentially sensitive attributes that have been added to the base model later):
```py
class UserInput(BaseModel):
    username: str
    fullname: str
    sensitiveInformation: str
    password: str


class UserOutput(UserInput):
    ...

    @model_serializer
    def model_serialize(self):
        return self.model_copy(include={'username', 'fullname'})
```

see https://docs.pydantic.dev/2.7/api/functional_serializers/#pydantic.functional_serializers.model_serializer

**Reply by mvariola** [reply]
FYI - `model_copy` does not have `include` param.

**Comment by tteguayco** [comment]
I'm surprised Pydantic doesn't provide a method to exclude fields in an inherited model out of the box.

How can this be raised as a feature request?

Relates to: https://github.com/pydantic/pydantic/issues/9573

**Reply by davidjb99** [reply]
@tteguayco does the answer posted here not work for you? 

https://github.com/pydantic/pydantic/discussions/6699#discussioncomment-8642547

**Comment by ShravanSunder** [comment]
i'm putting this in.  Its a way to create dynamic models from an a parent model that i'm using.  https://github.com/pydantic/pydantic/issues/9573#issuecomment-2196698767

**Comment by sudonorm** [comment]
`Pydantic v2`
This really had me pulling out my hair because though using `SkipJsonSchema` works as it should, the fields themselves were still in there, and this was throwing an error in a FastAPI application using the schema and preventing the Swagger UI from loading.

Not sure if 100% safe to do the below (although I don't see any downside or security issue), but I figured we can just use the `create_model()` method, since it returns a class and can be inherited by the class we are creating. 

If we want just a copy of the Parent class, then we can comment out the fields_to_include part `**{"b": (str, ...),}`. If we have no `model_config` for the `Child` class, we can use a `pass`.

For my use case, I had a Schema I did not want to repeat, but from which I wanted to remove some fields because I wanted to use it as an input for a FastAPI endpoint. The below code worked perfectly.

``` python

from pydantic import BaseModel, create_model
from pydantic.json_schema import SkipJsonSchema

class Parent(BaseModel):
    a: int = 1
    b: str = 'superstr'

class Child(create_model(
        "Parent",
        **{"b": (str, ...),} ### only include the fields we want, in this case, b
    )  
):
    model_config = {"extra": "forbid"} ## we can add extra configs to the new class
    # pass

# class Child(Parent):
#     a: SkipJsonSchema[int] = Field(default=1, exclude=True)
```

**Comment by superjgc** [comment]
pydantic==2.9.2, rewrite metaclass

```python
from pydantic import BaseModel
from pydantic._internal._model_construction import ModelMetaclass


class ModelMeta(ModelMetaclass):
    def __new__(cls, name, bases, namespace, **kwargs):
        epf = set(namespace.get("__exclude_parent_fields__", set()) or set())
        parent = bases[0]  # 至少会有BaseModel
        for base in parent.__mro__:
            if base == BaseModel:
                break
            base.__annotations__ = {
                k: v for k, v in base.__annotations__.items() if k not in epf
            }
            base.__pydantic_fields_set__ = {
                k for k in base.__annotations__.keys() if k not in epf
            }
            new_validators = {}
            validators = getattr(base, "__pydantic_decorators__", None)
            if validators:
                for (
                    func_name,
                    validator_decorator,
                ) in validators.field_validators.items():
                    if set(validator_decorator.info.fields) & epf:
                        continue
                    new_validators[func_name] = validator_decorator
                validators.field_validators = new_validators
        return super().__new__(cls, name, bases, namespace, **kwargs)


class Parent(BaseModel):
    key: str
    value: str


class Child(Parent, metaclass=ModelMeta):
    __exclude_parent_fields__ = ["value"]


if __name__ == "__main__":
    _parent = Parent(key="key", value="value")
    print(_parent)  # key='key' value='value'
    print(_parent.model_dump())  # {'key': 'key', 'value': 'value'}
    print(_parent.model_dump_json())  # {"key":"key","value":"value"}
    print(_parent.model_fields)
    # {'key': FieldInfo(annotation=str, required=True), 'value': FieldInfo(annotation=str, required=True)}

    _child = Child(key="child_key", value="child_value")
    print(_child)  # key='child_key'
    print(_child.model_dump())  # {'key': 'child_key'}
    print(_child.model_dump_json())  # {"key":"child_key"}
    print(_child.model_fields)  # {'key': FieldInfo(annotation=str, required=True)}

```

**Reply by superjgc** [reply]
Child(key="child_key").key -> this is ok
Child(key="child_key").value  or Child(key="child_key", value="child_value").value -> raise AttributeError

**Reply by skeppern** [reply]
Thank you for this. This saved me a lot of frustration and time and works like a charm.

**Reply by almabud-cogwheel** [reply]
Superb!! thanks for this.

**Comment by fbuccioni** [comment]
Based on @superjgc's answer, I suggest a factory function for this proposal, In fact, this will ensure that when exclude fields will be only from that parent class.

```python
from typing import Tuple
from pydantic import BaseModel
from pydantic._internal._model_construction import ModelMetaclass

def ExcludeFieldsModel(Model: BaseModel, *exclude_fields: Tuple[str]):
    if len(exclude_fields) < 1:
        raise ValueError("exclude_fields must be at least one field")

    class ExcludeFieldsMeta(ModelMetaclass):
        def __new__(cls, name, bases, namespace, **kwargs):
            epf = set(exclude_fields)

            if not len(bases):
                return super().__new__(cls, name, bases, namespace, **kwargs)
            
            parent = bases[0]
            for base in parent.__mro__:
                if base == BaseModel:
                    break
                if base != Model:
                    continue
                base.__annotations__ = {
                    k: v for k, v in base.__annotations__.items() if k not in epf
                }
                base.__pydantic_fields_set__ = {
                    k for k in base.__annotations__.keys() if k not in epf
                }
                new_validators = {}
                validators = getattr(base, "__pydantic_decorators__", None)
                if validators:
                    for (
                        func_name,
                        validator_decorator,
                    ) in validators.field_validators.items():
                        if set(validator_decorator.info.fields) & epf:
                            continue
                        new_validators[func_name] = validator_decorator
                    validators.field_validators = new_validators
            return super().__new__(cls, name, bases, namespace, **kwargs)

    class ExcludingFieldsModel(Model, metaclass=ExcludeFieldsMeta):
        pass

    return ExcludingFieldsModel
        

class Parent(BaseModel):
    key: str
    value: str


class Child(ExcludeFieldsModel(Parent, 'value')):
    pass


if __name__ == "__main__":
    _parent = Parent(key="key", value="value")
    print(_parent)  # key='key' value='value'
    print(_parent.model_dump())  # {'key': 'key', 'value': 'value'}
    print(_parent.model_dump_json())  # {"key":"key","value":"value"}
    print(_parent.model_fields)
    # {'key': FieldInfo(annotation=str, required=True), 'value': FieldInfo(annotation=str, required=True)}

    _child = Child(key="child_key", value="child_value")
    print(_child)  # key='child_key'
    print(_child.model_dump())  # {'key': 'child_key'}
    print(_child.model_dump_json())  # {"key":"child_key"}
    print(_child.model_fields)  # {'key': FieldInfo(annotation=str, required=True)}
```

**Comment by lmgarret** [comment]
I tried several of the options here but there was always one of my constraints that was not fulfilled:
 - the field should not appear in the json schema
 - mypy must be happy when instantiating the model, i.e. the field should at least become optional
 - pylance (vscode) must be happy when instantiating the model, , i.e. the field should at least become optional
 - dumping the model should exclude the field
 
 Here is what I ended up with, inspired by [this stackoverflow answer](https://stackoverflow.com/a/79176923):
 ```python
 from typing import Annotated, Any

from pydantic import AfterValidator, BaseModel, Field
from pydantic.json_schema import SkipJsonSchema

ExcludedField = SkipJsonSchema[
    Annotated[Any, Field(default=None, exclude=True), AfterValidator(lambda s: None)]
]


class ParentModel(BaseModel):
    a: int
    b: str
    c: int = 3


class ChildModel(ParentModel):
    a: ExcludedField = None  # Note: this assignation to None is important to please Pylance when constructing a ChildModel


child_a = ChildModel(b="test")
child_b = ChildModel(a=1, b="test")
print(child_a)
# a=None b='test' c=3
print(child_b)
# a=None b='test' c=3
print(child_a.model_dump(exclude_unset=True))
# {'b': 'test'}
print(child_b.model_dump(exclude_unset=True))
# {'b': 'test'}
print(child_a.model_dump())
# {'b': 'test', 'c': 3}
print(child_b.model_dump())
# {'b': 'test', 'c': 3}
```
